File: 1814-jump-game-vi/solution.cpp
#include <iostream>
#include <deque>

using namespace std;

class Solution {
public:
    int maxResult(vector<int>& nums, int k) {
        deque<int> queue;
        for (int i = 0; i < nums.size(); i++) {
            while (!queue.empty() && queue.front() < i - k) {
                queue.pop_front();
            }
            nums[i] += (!queue.empty() ? nums[queue.front()] : 0);
            while (!queue.empty() && nums[i] >= nums[queue.back()]) {
                queue.pop_back();
            }
            queue.push_back(i);
        }
        return nums.back();
    }
};

File: 1814-jump-game-vi/solution.py
from collections import deque


class Solution:
    def maxResult(self, nums: List[int], k: int) -> int:
        queue = deque([0])
        
        for i in range(1, len(nums)):
            while queue and queue[0] < i - k:
                queue.popleft()
            nums[i] += nums[queue[0]]
            while queue and nums[i] >= nums[queue[-1]]:
                queue.pop()
            queue.append(i)
        return nums[-1]
